src/include/type/type.h

```diff
@@ -14,6 +14,7 @@
 
 #include <cstdint>
 #include <string>
+#include <functional>
```

Comment:
The std::hash (http://en.cppreference.com/w/cpp/utility/hash) in line [35](https://github.com/AngLi-Leon/peloton/blob/13fa99bf672990ce891dde92a135e31d7dfd3e61/src/include/type/type.h#L35) need this.

If I don't explicitly include it here (under Mac), it will link to another lib and get error
```
/Users/angli/Documents/peloton_mac/peloton/src/include/type/type.h:35:14: error: implicit instantiation of undefined template 'std::__1::hash<unsigned int>'
      return std::hash<uint32_t>{}(static_cast<uint32_t>(type_id));
             ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:3171:29: note: template is declared here
template <class _Tp> struct hash;
                            ^
 
```